ui/src/components/FacetsGrid.js

```diff
@@ -1,26 +1,98 @@
 import './FacetsGrid.css';
-
-import React from 'react';
 import FacetCard from "./FacetCard";
+import ApiClient from "../api/src/ApiClient";
+import FacetsApi from "../api/src/api/FacetsApi";
+
+import React, { Component } from 'react';
 import {GridList, GridTile} from "material-ui/GridList";
 
-function FacetsGrid(props) {
-    const facetsResponse = props.facetsResponse;
-    const facetsList = facetsResponse.facets.map((facet) =>
-        <GridTile key={facet.name}>
-            <FacetCard facet={facet} count={facetsResponse.count} className='facetCard'/>
-        </GridTile>
-    );
-    return (
-        <GridList
-            className="gridList"
-            cols={3}
-            cellHeight="auto"
-            padding={1}
-        >
-            {facetsList}
-        </GridList>
-    );
+class FacetsGrid extends Component {
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            facetsResponse: props.facetsResponse
+        };
+        this.searchFilters = new Map();
+        let apiClient = new ApiClient();
+        apiClient.basePath = '/api';
+        this.api = new FacetsApi(apiClient);
+        this.callback = function(error, data) {
+            if (error) {
+                console.error(error);
+                // TODO(alanhwang): Redirect to an error page
+            } else {
+                this.setState({facetsResponse: data});
+            }
+        }.bind(this);
+        this.updateFacets = this.updateFacets.bind(this);
+    }
+
+    render() {
+        const facetsList = this.state.facetsResponse.facets.map((facet) =>
+            <GridTile key={facet.name}>
+                <FacetCard
+                    facet={facet}
+                    count={this.state.facetsResponse.count}
+                    updateFacets={this.updateFacets}
+                    className='facetCard'/>
+            </GridTile>
+        );
+        return (
+            <GridList
+                className="gridList"
+                cols={3}
+                cellHeight="auto"
+                padding={1}
+            >
+                {facetsList}
+            </GridList>
+        );
+    }
+
+    /**
+     * Updates the selection for a single facet value and refreshes the facets data from the server.
+     * @param facetName string containing the name of the facet corresponding to this value
+     * @param facetValue string containing the name of this facet value
+     * @param isSelected bool indicating whether this facetValue should be added to or removed from the query
+     * */
+    updateFacets(facetName, facetValue, isSelected) {
+        let currFacetVals = this.searchFilters.get(facetName);
+        if (isSelected) {
+            // Add facetValue to the list of filters for facetName
+            if (currFacetVals === undefined) {
```

Comment:
I would expect either
```
            if (currFacetVals === undefined) {
                this.searchFilters.set(facetName, [facetValue]);
            } else {
                currFacetVals.push(facetValue);
            }
```
or
````
            if (currFacetVals === undefined) {
                this.searchFilters.set(facetName, [facetValue]);
            } else if (!this.searchFilters.get(facetName).includes(facetValue)) {
                currFacetVals.push(facetValue);
            } else {
                 show an error
            }
```

Reply:
Actually the former is fine because removeFacet will always iterate through the whole list anyway to remove potentially multiple copies of the same Facet in the worst case of us getting to this bad state.

Reply:
Done